bindings/go/src/fdb/tuple/tuple.go

```diff
@@ -106,109 +115,128 @@ func adjustFloatBytes(b []byte, encode bool) {
 	}
 }
 
-func encodeBytes(buf *bytes.Buffer, code byte, b []byte) {
-	buf.WriteByte(code)
-	buf.Write(bytes.Replace(b, []byte{0x00}, []byte{0x00, 0xFF}, -1))
-	buf.WriteByte(0x00)
+type packer struct {
+	buf []byte
 }
 
-func bisectLeft(u uint64) int {
-	var n int
-	for sizeLimits[n] < u {
-		n++
+func (p *packer) putbyte(b byte) {
+	p.buf = append(p.buf, b)
+}
+
+func (p *packer) putbytes(b []byte) {
+	p.buf = append(p.buf, b...)
+}
+
+func (p *packer) putbytesNil(b []byte, i int) {
+	for i >= 0 {
+		p.putbytes(b[:i+1])
+		p.putbyte(0xFF)
+		b = b[i+1:]
+		i = bytes.IndexByte(b, 0x00)
 	}
-	return n
+	p.putbytes(b)
 }
 
-func encodeInt(buf *bytes.Buffer, i int64) {
+func (p *packer) encodeBytes(code byte, b []byte) {
+	p.putbyte(code)
+	if i := bytes.IndexByte(b, 0x00); i >= 0 {
+		p.putbytesNil(b, i)
+	} else {
+		p.putbytes(b)
+	}
+	p.putbyte(0x00)
+}
+
+func (p *packer) encodeInt(i int64) {
 	if i == 0 {
-		buf.WriteByte(0x14)
+		p.putbyte(0x14)
 		return
 	}
 
 	var n int
-	var ibuf bytes.Buffer
+	var scratch [8]byte
 
 	switch {
 	case i > 0:
 		n = bisectLeft(uint64(i))
-		buf.WriteByte(byte(intZeroCode + n))
-		binary.Write(&ibuf, binary.BigEndian, i)
+		p.putbyte(byte(intZeroCode + n))
+		binary.BigEndian.PutUint64(scratch[:], uint64(i))
 	case i < 0:
 		n = bisectLeft(uint64(-i))
-		buf.WriteByte(byte(0x14 - n))
-		binary.Write(&ibuf, binary.BigEndian, int64(sizeLimits[n])+i)
+		p.putbyte(byte(0x14 - n))
+		ii := int64(sizeLimits[n]) + i
+		binary.BigEndian.PutUint64(scratch[:], uint64(ii))
```

Comment:
Hm, this is nit-picking (sorry), but is there a better name that `ii` for that variable? If it weren't obnoxious, it might make sense to call it, like `offsetEncodedI`, but that's somewhat of a mouthful.

Reply:
Fixed!